### PR TITLE
Move away from using the instance id and to fetch datasets

### PR DESCRIPTION
**Pull request by Matt-Guest**
### What

Use dataset_id, edition and version to create a filter job. This change is needed as the filter api can no longer call the instance endpoints on the dataset api

### How to review

review the code, run the unit tests. 

### Who can review

Anyone

### CODE REVIEW

File: api/filters.go [hunk 1]
```diff
@@ -779,11 +775,11 @@ func setErrorCode(w http.ResponseWriter, err error, typ ...string) {
 	case "Instance not found":
```
Comment (CarlHembrough) [inline api/filters.go]: Should this error message be changed from instance to version?

File: models/models.go [hunk 2]
```diff
@@ -252,3 +271,23 @@ func CreateDimensionOptions(reader io.Reader) ([]string, error) {
...
+	}
+
+	var filter NewFilter
+	err = json.Unmarshal(bytes, &filter)
+	if err != nil {
+		return nil, ErrorParsingBody
+	}
+
+	// This should be the last check before returning filter
+	if len(bytes) == 2 {
```
Comment (CarlHembrough) [inline models/models.go]: Why 2 bytes? consider declaring a descriptive variable name or adding a comment

Comment (Matt-Guest) [reply models/models.go]: See comment.

File: api/filters.go (hunk 1 above)
Comment (Matt-Guest) [reply api/filters.go]: Done

**Comment by Matt-Guest** [comment]
@CarlHembrough Nathan got some additional changes he going to push up as well

File: api/dataset.go [hunk 3]
```diff
@@ -177,17 +189,17 @@ func (api *DatasetAPI) callDatasetAPI(ctx context.Context, method, path string,
...
-		return nil, resp.StatusCode, ErrUnexpectedStatusCode
-	}
-
 	jsonBody, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		log.ErrorC("failed to read body from dataset api", err, logData)
 		return nil, resp.StatusCode, err
 	}
 
+	logData["httpCode"] = resp.StatusCode
```
Comment (CarlHembrough) [inline api/dataset.go]: The check of the response code has been moved from before reading the response body to after. Don't we want to check the response code first?

File: api/filters.go [hunk 4]
```diff
@@ -337,10 +400,18 @@ func (api *FilterAPI) addFilterBlueprintDimensionOption(w http.ResponseWriter, r
 		Name:    addDimensionOption.Name,
 		Options: []string{addDimensionOption.Option},
 	}
+	logData["dimension_options"] = addDimensionOptions
 
 	if err = api.checkNewFilterDimension(r.Context(), addDimensionOptions, filterBlueprint.Dataset); err != nil {
-		log.Error(err, nil)
+		log.ErrorC("got here", err, logData)
```
Comment (CarlHembrough) [inline api/filters.go]: Add more meaningful context here 😄

File: api/filters.go [hunk 5]
```diff
@@ -337,10 +400,18 @@ func (api *FilterAPI) addFilterBlueprintDimensionOption(w http.ResponseWriter, r
...
+		if err == ErrVersionNotFound {
+			log.Debug("nathan here", nil)
```
Comment (CarlHembrough) [inline api/filters.go]: and here 😂

File: api/filters.go [hunk 6]
```diff
@@ -31,28 +31,32 @@ var (
...
 func (api *FilterAPI) addFilterBlueprint(w http.ResponseWriter, r *http.Request) {
 	submitted := r.FormValue("submitted")
+	logData := log.Data{"func": "addFilterBlueprint", "submitted": submitted}
+	log.Info("create filter blueprint", logData)
 
 	newFilter := models.Filter{}
 	filterParameters, err := models.CreateNewFilter(r.Body)
 	if err != nil {
-		log.Error(err, nil)
+		log.Error(err, logData)
```
Comment (CarlHembrough) [inline api/filters.go]: Consider adding context messages to logs using `log.ErrorC`

File: api/filters.go [hunk 7]
```diff
@@ -286,46 +332,63 @@ func (api *FilterAPI) getFilterBlueprintDimensionOptions(w http.ResponseWriter,
...
+	log.Info("got filtered blueprint", logData)
 }
 
 func (api *FilterAPI) addFilterBlueprintDimensionOption(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
+	filterID := vars["filter_blueprint_id"]
+	name := vars["name"]
+	option := vars["option"]
+	logData := log.Data{
+		"func":                "addFilterBlueprintDimensionOption",
```
Comment (CarlHembrough) [inline api/filters.go]: I wonder if its necessary to have the func names in logs. Having the first log.info in the function provides the required context to determine the function. Also I can imagine them becoming stale if the function name changes.

File: api/filters.go [hunk 8]
```diff
@@ -574,57 +682,75 @@ func (api *FilterAPI) removeFilterBlueprintDimensionOption(w http.ResponseWriter
...
+	log.Info("get filter output preview", logData)
+
 	var limit = 20
 	var err error
 	if requestedLimit != "" {
 		limit, err = strconv.Atoi(requestedLimit)
 		if err != nil {
-			log.Error(errors.New("requested limit is not a number"), log.Data{"filter_output_id": filterID, "limit": limit})
+			logData["requested_limit"] = requestedLimit
+			log.Error(errors.New("requested limit is not a number"), logData)
```
Comment (CarlHembrough) [inline api/filters.go]: Consider declaring the error at package level

File: models/models.go [hunk 9]
```diff
@@ -230,6 +230,30 @@ func (filter *Filter) ValidateFilterOutputUpdate() error {
...
+// ValidateFilterBlueprintUpdate checks the content of the filter structure for
+// changes against the dataset
+func ValidateFilterBlueprintUpdate(filter *Filter) error {
+
+	// Only events and dataset version can be updated, any attempt to update the
+	// dataset id or edition will result in an error of bad request
+
+	var forbiddenFields []string
+
+	if filter.Dataset.ID != "" {
```
Comment (CarlHembrough) [inline models/models.go]: Should we also do checks for nil here?

File: api/dataset.go (hunk 3 above)
Comment (nshumoogum) [reply api/dataset.go]: yes

File: api/filters.go (hunk 7 above)
Comment (nshumoogum) [reply api/filters.go]: Good point, it was helpful whilst developing but maybe not so much in kibana

File: api/filters.go [hunk 10]
```diff
@@ -356,98 +420,106 @@ func (api *FilterAPI) addFilterBlueprintDimensionOption(w http.ResponseWriter, r
...
-		log.Error(err, log.Data{"filter_blueprint": filter, "filter_blueprint_id": filterID})
+		log.Error(err, logData)
 		setErrorCode(w, err)
 		return
 	}
+	logData["current_filter"] = currentFilter
 
-	newFilter := currentFilter
-	if filter.InstanceID != "" {
+	newFilter, VersionHasChanged := createNewFilter(filter, currentFilter)
```
Comment (eldeal) [inline api/filters.go]: Should this variable really be exported?

Comment (nshumoogum) [reply api/filters.go]: Nope, just a typo, I will remove it

**Pull request merged**